File: src/main/ts/com/jcloisterzone/game/capability/RiverCapability.ts
import { HashSet, type Set } from "../../../../io/vavr/Set.js";
import { List, Vector } from "../../../../io/vavr/SeqTypes.js";
import { EdgeType } from "../../board/EdgeType.js";
import { Location } from "../../board/Location.js";
import { Position } from "../../board/Position.js";
import type { PlacementOption } from "../../board/PlacementOption.js";
import { Rotation } from "../../board/Rotation.js";
import type { Tile } from "../../board/Tile.js";
import { TileGroup } from "../../board/TileGroup.js";
import type { TilePack } from "../../board/TilePack.js";
import { River } from "../../feature/River.js";
import type { RandomGenerator } from "../../random/RandomGenerator.js";
import { PlaceTile } from "../../reducers/PlaceTile.js";
import { Capability } from "../Capability.js";
import { Flag } from "../state/Flag.js";
import type { GameState } from "../state/GameState.js";
import type { PlacedTile } from "../state/PlacedTile.js";
import { DragonCapability } from "./DragonCapability.js";

/** Name of the group holding the single volcano (or, lacking one, random) lake forced to be the
 *  last river tile drawn — so the dragon is summoned exactly when the river finishes. */
const RIVER_LAKE_VOLCANO = "river-lake-volcano";

/** The River — a separate tile group drawn first, with placement rules that forbid
 *  U-turns and same-direction river junctions. Port of RiverCapability. */
export class RiverCapability extends Capability<void> {
  override onStartGame(state: GameState, random: RandomGenerator): GameState {
    return state.mapTilePack((pack: TilePack) => {
      pack = pack.deactivateGroup("default");
      pack = pack.deactivateGroup("river-lake");
      pack = pack.mapGroup("river", (g) => g.setSuccessiveGroup("river-lake"));
      pack = pack.mapGroup("river-lake", (g) => g.setSuccessiveGroup("default"));

      if (pack.hasGroup("river-fork")) {
        const riverForks = pack.getGroup("river-fork")!;
        let riverForksLocationsCount = 0;
        for (const t of riverForks.getTiles()) {
          for (const loc of this.getRiverLocations(t, Rotation.R0)) {
            riverForksLocationsCount += loc.splitToSides().size();
          }
        }
        pack = pack.mapGroup("river-fork", (g) => g.setSuccessiveGroup("river"));
        pack = pack.deactivateGroup("river");

        // lakes needed = 1 initial open branch + each fork's (ends - 2):
        // a fork uses 1 existing open end to connect and adds (ends - 1) new ones, net (ends - 2).
        const branches = 1 + (riverForksLocationsCount - 2 * riverForks.getTiles().size());
        const ends = pack.getGroupSize("river-lake");
        if (branches !== ends) {
          const riverLakes = pack.getGroup("river-lake")!.getTiles();
          const trimmed = this.adjustRandomTiles(riverLakes, branches - ends, random);
          pack = pack.mapGroup("river-lake", (g) => g.setTiles(trimmed));
        }
      }
      // Pull ONE lake — a volcano lake if any (so the dragon is summoned exactly when the river
      // finishes), otherwise any lake — into its own group drawn LAST. The remaining lakes stay in
      // the river-lake group, drawn (via the successive chain) AFTER all river tiles and before the
      // final lake. Chain: river → river-lake → river-lake-volcano → default.
      if (pack.hasGroup("river-lake") && pack.getGroupSize("river-lake") > 0) {
        const lakes = pack.getGroup("river-lake")!.getTiles();
        const volcanoLakes = lakes.filter((t) => this.isVolcanoTile(t)) as Vector<Tile>;
        const candidates = (volcanoLakes.isEmpty() ? lakes : volcanoLakes) as Vector<Tile>;
        const chosen = candidates.get(random.getNextInt(candidates.size()));
        const remaining = lakes.removeFirst((t) => t === chosen) as Vector<Tile>;
        if (remaining.isEmpty()) {
          // the chosen lake was the only one — river flows straight into the volcano-lake group
          // (an empty river-lake group would never trigger its successive).
          pack = pack.removeGroup("river-lake");
          pack = pack.mapGroup("river", (g) => g.setSuccessiveGroup(RIVER_LAKE_VOLCANO));
        } else {
          // the other lakes stay in river-lake (chain-activated when river drains), flowing into
          // the volcano-lake group instead of straight to default.
          pack = pack.mapGroup("river-lake", (g) =>
            g.setTiles(remaining).setSuccessiveGroup(RIVER_LAKE_VOLCANO),
          );
        }
        pack = pack.updateGroup(new TileGroup(RIVER_LAKE_VOLCANO, Vector.of(chosen) as Vector<Tile>, false, "default"));
      }
      pack = pack.removeGroup("river-spring"); // remove unused springs
      return pack;
    });
  }

  override onTilePlaced(state: GameState, placedTile: PlacedTile): GameState {
    if (state.getPlacedTiles().size() > 1) {
      // first tile already placed (the river-fork if any) → mix other forks in
      state = state.mapTilePack((pack) => pack.activateGroup("river"));
    }
    // The volcano lake is forced to be the last river tile; whoever places it takes another turn
    // (so they play the first tile of the now-active main deck the dragon just joined).
    if (placedTile.getTile().hasModifier(DragonCapability.VOLCANO)) {
      state = state.addFlag(Flag.RIVER_VOLCANO_DOUBLE_TURN);
    }
    return state;
  }

  private isVolcanoTile(tile: Tile): boolean {
    return tile.hasModifier(DragonCapability.VOLCANO);
  }

  override isTilePlacementAllowed(state: GameState, tile: Tile, placement: PlacementOption): boolean {
    const pos = placement.getPosition();
    const rot = placement.getRotation();
    const riverLocations = this.getRiverLocations(tile, rot);
    if (riverLocations.size() === 0) return true;

    let foundValidRiverPlacement = false;
    for (const riverLoc of riverLocations) {
      const sides = riverLoc.splitToSides();
      const openSides = sides.filter((side) => !this.isConnectedToPlacedRiver(state, pos, side));
      const openSidesSize = openSides.size();
      if (sides.size() !== openSidesSize) {
        foundValidRiverPlacement = true;
        if (
          openSides.filter((side) => this.isContinuationFree(state, pos, side, tile, rot)).size() !==
          openSidesSize
        ) {
          return false;
        }
      }
    }
    return foundValidRiverPlacement;
  }

  /** The unioned river side-location(s) of a tile (rotated), or empty if not a river tile. */
  private getRiverLocations(tile: Tile, rot: Rotation): List<Location> {
    const sideLocs: Location[] = [];
    for (const t of tile.getInitialFeatures()) {
      if (!(t._2 instanceof River)) continue;
      const loc = t._1.getLocation()!.rotateCW(rot);
      for (const s of loc.splitToSides()) sideLocs.push(s);
    }
    if (sideLocs.length === 0) return List.empty<Location>();
    let u = sideLocs[0];
    for (let i = 1; i < sideLocs.length; i++) u = u.union(sideLocs[i]);
    return List.of(u);
  }

  private isConnectedToPlacedRiver(state: GameState, pos: Position, side: Location): boolean {
    return state.getPlacedTiles().containsKey(pos.add(side));
  }

  private isContinuationFree(
    state: GameState,
    pos: Position,
    side: Location,
    tile: Tile,
    rot: Rotation,
  ): boolean {
    const adjPos = pos.add(side);
    const adjPos2 = adjPos.add(side);
    const reserved = [
      adjPos.add(side.prev()),
      adjPos.add(side.next()),
      adjPos2,
      adjPos2.add(side.prev()),
      adjPos2.add(side.next()),
    ];
    for (const p of reserved) if (state.getPlacedTiles().containsKey(p)) return false; // U-turn

    const placed = new PlaceTile(tile, pos, rot).apply(state);
    const placedTiles = placed.getPlacedTiles();
    let minX = 0,
      maxX = 0,
      minY = 0,
      maxY = 0;
    for (const p of placedTiles.keySet()) {
      if (p.x < minX) minX = p.x;
      if (p.y < minY) minY = p.y;
      if (p.x > maxX) maxX = p.x;
      if (p.y > maxY) maxY = p.y;
    }

    const check: Position[] = [];
    let edgeI = -1;
    if (side === Location.N) {
      edgeI = 0;
      for (let y = minY; y < pos.y; y++)
        check.push(new Position(pos.x - 1, y), new Position(pos.x, y), new Position(pos.x + 1, y));
    } else if (side === Location.E) {
      edgeI = 1;
      for (let x = pos.x + 1; x <= maxX; x++)
        check.push(new Position(x, pos.y - 1), new Position(x, pos.y), new Position(x, pos.y + 1));
    } else if (side === Location.S) {
      edgeI = 2;
      for (let y = pos.y + 1; y <= maxY; y++)
        check.push(new Position(pos.x - 1, y), new Position(pos.x, y), new Position(pos.x + 1, y));
    } else if (side === Location.W) {
      edgeI = 3;
      for (let x = minX; x < pos.x; x++)
        check.push(new Position(x, pos.y - 1), new Position(x, pos.y), new Position(x, pos.y + 1));
    }
    if (check.length > 0 && check.some((p) => placedTiles.containsKey(p))) return false; // tile in straight path

    const edgePositions: Position[] = [];
    for (const placement of placed.getAvailablePlacements()) {
      const edges = placement._2.getEdges();
      if (edges[(edgeI + 2) % 4] === EdgeType.RIVER) edgePositions.push(placement._1);
    }
    if (edgePositions.length >= 2) {
      for (let i = 0; i < edgePositions.length; i++) {
        for (let j = i + 1; j < edgePositions.length; j++) {
          const p1 = edgePositions[i];
          const p2 = edgePositions[j];
          const consecutive =
            edgeI === 0 || edgeI === 2 ? Math.abs(p1.x - p2.x) === 1 : Math.abs(p1.y - p2.y) === 1;
          if (consecutive) return false; // river junction following the curve
        }
      }
    }
    return true;
  }

  private adjustRandomTiles(tiles: Vector<Tile>, count: number, random: RandomGenerator): Vector<Tile> {
    if (count > 0) {
      // Add `count` extra lake ends by duplicating random lake tiles. Java does NOT
      // exclude the protected volcano here (the duplicate is only a draw-pile entry,
      // not a second physical tile) and it MUST consume one getNextInt per added end
      // to keep the RNG in lock-step with the Java engine.
      return tiles.appendAll(
        Vector.range(0, count).map(() => tiles.get(random.getNextInt(tiles.size()))),
      ) as Vector<Tile>;
    }
    // Never remove a volcano lake (detected by modifier, not by tile id).
    let pool = Vector.range(0, tiles.size()).filter(
      (idx) => !this.isVolcanoTile(tiles.get(idx)),
    ) as Vector<number>;
    const toRemove = Math.min(-count, pool.size());
    let removeIdx: Set<number> = HashSet.empty<number>();
    for (let i = 0; i < toRemove; i++) {
      const p = random.getNextInt(pool.size());
      removeIdx = removeIdx.add(pool.get(p));
      pool = pool.removeAt(p) as Vector<number>;
    }
    return tiles
      .zipWithIndex()
      .filter((t) => !removeIdx.contains(t._2))
      .map((t) => t._1) as Vector<Tile>;
  }
}

Capability.register(RiverCapability);

File: src/main/java/com/jcloisterzone/game/capability/RiverCapability.java
package com.jcloisterzone.game.capability;

import com.jcloisterzone.board.*;
import com.jcloisterzone.feature.River;
import com.jcloisterzone.game.Capability;
import com.jcloisterzone.game.state.Flag;
import com.jcloisterzone.game.state.GameState;
import com.jcloisterzone.game.state.PlacedTile;
import com.jcloisterzone.random.RandomGenerator;
import com.jcloisterzone.reducers.PlaceTile;
import io.vavr.Predicates;
import io.vavr.Tuple2;
import io.vavr.collection.HashSet;
import io.vavr.collection.LinkedHashMap;
import io.vavr.collection.List;
import io.vavr.collection.Set;
import io.vavr.collection.Stream;
import io.vavr.collection.Vector;

public class RiverCapability extends Capability<Void> {

	private static final long serialVersionUID = 1L;

	/** Group holding the single volcano (or, lacking one, random) lake forced to be the last
	 *  river tile drawn — so the dragon is summoned exactly when the river finishes. */
	private static final String RIVER_LAKE_VOLCANO = "river-lake-volcano";

    @Override
    public GameState onStartGame(GameState state, RandomGenerator random) {
        state = state.mapTilePack(pack -> {
            pack = pack.deactivateGroup("default");
            pack = pack.deactivateGroup("river-lake");
            pack = pack.mapGroup("river", g -> g.setSuccessiveGroup("river-lake"));
            pack = pack.mapGroup("river-lake", g -> g.setSuccessiveGroup("default"));
            
            if (pack.hasGroup("river-fork")) {
                TileGroup riverForks = pack.getGroup("river-fork");
                int riverForksLocationsCount = riverForks.getTiles().map(t -> getRiverLocations(t, Rotation.R0)).flatMap(list -> list.flatMap(loc -> loc.splitToSides())).size();

                pack = pack.mapGroup("river-fork", g -> g.setSuccessiveGroup("river"));
                pack = pack.deactivateGroup("river");

                // lakes needed = 1 initial open branch + each fork's (ends - 2):
                // a fork uses 1 existing open end to connect and adds (ends - 1) new ones, net (ends - 2).
                int branches = 1 + (riverForksLocationsCount - 2 * riverForks.getTiles().size());
                int ends = pack.getGroupSize("river-lake");
                if (branches != ends) {
                    Vector<Tile> riverLakes = pack.getGroup("river-lake").getTiles();
                    final Vector<Tile> trimmed = adjustRandomTiles(riverLakes, branches - ends, random);
                    pack = pack.mapGroup("river-lake", g -> g.setTiles(trimmed));
                }
            }
            // Pull ONE lake — a volcano lake if any (so the dragon is summoned exactly when the
            // river finishes), otherwise any lake — into its own group drawn LAST. The remaining
            // lakes stay in the river-lake group, drawn (via the successive chain) AFTER all river
            // tiles and before the final lake. Chain: river -> river-lake -> river-lake-volcano -> default.
            if (pack.hasGroup("river-lake") && pack.getGroupSize("river-lake") > 0) {
                Vector<Tile> lakes = pack.getGroup("river-lake").getTiles();
                Vector<Tile> volcanoLakes = lakes.filter(this::isVolcanoTile);
                Vector<Tile> candidates = volcanoLakes.isEmpty() ? lakes : volcanoLakes;
                final Tile chosen = candidates.get(random.getNextInt(candidates.size()));
                Vector<Tile> remaining = lakes.removeFirst(t -> t == chosen);
                if (remaining.isEmpty()) {
                    // the chosen lake was the only one — river flows straight into the volcano-lake
                    // group (an empty river-lake group would never trigger its successive).
                    pack = pack.removeGroup("river-lake");
                    pack = pack.mapGroup("river", g -> g.setSuccessiveGroup(RIVER_LAKE_VOLCANO));
                } else {
                    // the other lakes stay in river-lake (chain-activated when river drains),
                    // flowing into the volcano-lake group instead of straight to default.
                    pack = pack.mapGroup("river-lake", g ->
                        g.setTiles(remaining).setSuccessiveGroup(RIVER_LAKE_VOLCANO));
                }
                pack = pack.setGroups(pack.getGroups().put(RIVER_LAKE_VOLCANO,
                    new TileGroup(RIVER_LAKE_VOLCANO, Vector.of(chosen), false, "default")));
            }
            pack = pack.removeGroup("river-spring"); // remove unused springs
            return pack;
        });
        return state;
    }

    @Override
    public GameState onTilePlaced(GameState state, PlacedTile placedTile) {
    	if (state.getPlacedTiles().size()>1) {
    	    // Already placed first tile. It was first river-fork if exists
            // mix other forks between other river tiles (this is applied when multiple rivers are enabled
            state = state.mapTilePack(pack -> pack.activateGroup("river"));
        }
        // The volcano lake is forced to be the last river tile; whoever places it takes another
        // turn (so they play the first tile of the now-active main deck the dragon just joined).
        if (placedTile.getTile().hasModifier(DragonCapability.VOLCANO)) {
            state = state.setFlags(state.getFlags().add(Flag.RIVER_VOLCANO_DOUBLE_TURN));
        }
        return state;
    }

    private boolean isVolcanoTile(Tile tile) {
        return tile.hasModifier(DragonCapability.VOLCANO);
    }

    private boolean isConnectedToPlacedRiver(GameState state, Position pos, Location side) {
        Position adjPos = pos.add(side);
        return state.getPlacedTiles().containsKey(adjPos);
    }
    
    private List<Location> getRiverLocations(Tile tile, Rotation rot) {
    	List<Location> riverLocations = tile.getInitialFeatures()
            .filterValues(Predicates.instanceOf(River.class))
            .map(Tuple2::_1)
            .map(fp -> fp.getLocation().rotateCW(rot))
            .flatMap(riverLoc -> riverLoc.splitToSides())
            .reduceOption(Location::union)
            .map(loc -> List.of(loc))
            .getOrElse(List.of());

	    return riverLocations;
    }

    private boolean isContinuationFree(GameState state, Position pos, Location side, Tile tile, Rotation rot) {
        Position adjPos = pos.add(side);
        Position adjPos2 = adjPos.add(side);
        List<Position> reservedTiles = List.of(
            adjPos.add(side.prev()),
            adjPos.add(side.next()),
            adjPos2,
            adjPos2.add(side.prev()),
            adjPos2.add(side.next())
        );
        if (!reservedTiles.find(p -> state.getPlacedTiles().containsKey(p)).isEmpty()) {
        	// Found regular U-turn
        	return false;
        }
        
        GameState _state = (new PlaceTile(tile, pos, rot)).apply(state);

        LinkedHashMap<Position, PlacedTile> placedTiles = _state.getPlacedTiles();
        
        int minX = 0, maxX = 0, minY = 0, maxY = 0;

        for (Position p : _state.getPlacedTiles().keySet()) {
            int x = p.x;
            int y = p.y;

            if (x < minX) minX = x;
            if (y < minY) minY = y;
            if (x > maxX) maxX = x;
            if (y > maxY) maxY = y;
        }

    	List<Position> check = List.empty();
        
    	int edgeI = -1;

    	if (side.equals(Location.N)) {
    		edgeI = 0;
			for(int y = minY; y<pos.y; y++) {
	    		check = check.append(new Position(pos.x-1,y));
	    		check = check.append(new Position(pos.x,y));
	    		check = check.append(new Position(pos.x+1,y));
			}
    	} else if (side.equals(Location.E)) {
    		edgeI = 1;
			for(int x = pos.x+1; x<=maxX; x++) {
				check = check.append(new Position(x,pos.y-1));
				check = check.append(new Position(x,pos.y));
				check = check.append(new Position(x,pos.y+1));
			}
    	} else if (side.equals(Location.S)) {
    		edgeI = 2;
			for(int y = pos.y+1; y<=maxY; y++) {
	    		check = check.append(new Position(pos.x-1,y));
	    		check = check.append(new Position(pos.x,y));
	    		check = check.append(new Position(pos.x+1,y));
			}
    	} else if (side.equals(Location.W)) {
    		edgeI = 3;
			for(int x = minX; x<pos.x; x++) {
				check = check.append(new Position(x,pos.y-1));
				check = check.append(new Position(x,pos.y));
				check = check.append(new Position(x,pos.y+1));
			}
		}
		if (check.size()>0 && check.exists(p -> placedTiles.containsKey(p))) {
			// In straight direction of river branch is at least one tile, which can cause problem with River finishing
		    return false;
		}
		
        Stream<Tuple2<Position, EdgePattern>> availableUnfilteredPlacements = _state.getAvailablePlacements();
        
        List<Position> edgePositions = List.empty();

        for (Tuple2<Position, EdgePattern> placement : availableUnfilteredPlacements) {
            EdgeType[] edges = placement._2.getEdges();
            // We are checking opposite edge
            if (edges[(edgeI+2)%4] == EdgeType.RIVER) {
            	edgePositions = edgePositions.append(placement._1);
            }
        }
        boolean hasConsecutivePosition = false;
        	
        if (edgePositions.size()>=2) {

        	for (int i = 0; i < edgePositions.size(); i++) {
        	    Position p1 = edgePositions.get(i);
        	    for (int j = i + 1; j < edgePositions.size(); j++) {
        	        Position p2 = edgePositions.get(j);

        	        if (edgeI == 0 || edgeI == 2) {
        	        	if (Math.abs(p1.x - p2.x) == 1) { // consecutive x
        	        		hasConsecutivePosition = true;
        	        		break;
        	        	}
        	        } else {
        	        	if (Math.abs(p1.y - p2.y) == 1) { // consecutive y
        	        		hasConsecutivePosition = true;
        	        		break;
        	        	}
        	        }
        	    }
        	}
    	    if (hasConsecutivePosition) {
    	        // River junction following river curve. Two open neighbouring river branches with same direction
    	    	return false;
    	    }
        }
        // Possible future testings
        // TODO: Check possible collision with all unfinished river branch in straight lines
        return true;
    }

    @Override
    public boolean isTilePlacementAllowed(GameState state, Tile tile, PlacementOption placement) {
        Position pos = placement.getPosition();
        Rotation rot = placement.getRotation();
        List<Location> riverLocations = getRiverLocations(tile, rot);

        if (riverLocations.size() == 0) {
            return true;
        }

        boolean foundValidRiverPlacement = false;
        
        for (Location riverLoc : riverLocations) {
            List<Location> sides = riverLoc.splitToSides();
            List<Location> openSides = sides.filter(side -> !isConnectedToPlacedRiver(state, pos, side));
            List<Location> connectedSides = sides.filter(side -> isConnectedToPlacedRiver(state, pos, side));

            int openSidesSize = openSides.size();
            
            if (sides.size() != openSidesSize) {
            	foundValidRiverPlacement = true;

            	if (openSides.filter(side -> isContinuationFree(state, pos, side, tile, rot)).size() != openSidesSize) {
	            	return false;
	            }
            }
        }
        
        return foundValidRiverPlacement;
    }
    
    private Vector<Tile> adjustRandomTiles(Vector<Tile> tiles, int count, RandomGenerator random) {
        if (count > 0) {
            return tiles.appendAll(Vector.range(0, count)
                .map(i -> tiles.get(random.getNextInt(tiles.size()))));
        } else {
        	// Never remove a volcano lake (detected by modifier, not by tile id).
        	Vector<Integer> pool = Vector.range(0, tiles.size())
                .filter(idx -> !isVolcanoTile(tiles.get(idx)));
            int toRemove = Math.min(-count, pool.size());

            Set<Integer> removeIdx = HashSet.empty();
            for (int i = 0; i < toRemove; i++) {
                int p = random.getNextInt(pool.size());
                removeIdx = removeIdx.add(pool.get(p));
                pool = pool.removeAt(p);
            }
            final Set<Integer> finalRemoveIdx = removeIdx;

            return tiles.zipWithIndex()
                .filter(t -> !finalRemoveIdx.contains(t._2))
                .map(t -> t._1);
        }
    }
}